compiler/pipes/final-check.cpp

```diff
@@ -107,6 +109,97 @@ void check_to_array_debug_call(VertexAdaptor<op_func_call> call) {
+      auto key = rename.empty() ? field.local_name() : rename;
+      bool success = used_json_keys.emplace(key).second;
+      kphp_error_return(success, fmt_format("@kphp-json field '{}' met twice, first time in class '{}'", key, ancestor->name));
+    });
+  }
+}
+
+void check_json_tag_correctness(ClassPtr klass) noexcept {
+  klass->members.for_each([](const ClassMemberInstanceField &field) {
+    const auto type = tinf::get_type(field.var)->ptype();
+    if (field.kphp_json_tags && field.kphp_json_tags->has_array_as_hashmap()) {
+      kphp_error_return(type == tp_array,
+        fmt_format("@kphp-json array_as_hashmap tag is allowed only above array type, got above {} field", field.local_name()));
+    }
+    if (field.kphp_json_tags && field.kphp_json_tags->is_raw_string()) {
+      kphp_error_return(type == tp_string,
+        fmt_format("@kphp-json raw_string tag is allowed only above string type, got above {} field", field.local_name()));
+    }
+  });
+}
+
+void store_json_encoder(ClassPtr klass, ClassPtr encoder) noexcept {
+  if (klass->parent_class) {
+    store_json_encoder(klass->parent_class, encoder);
+  }
+  std::for_each(klass->implements.begin(), klass->implements.end(), [encoder](InterfacePtr i) {
+    store_json_encoder(i, encoder);
+  });
+
+  AutoLocker<Lockable *> lock{&(*klass)};
+  auto it = std::find(klass->json_encoders.begin(), klass->json_encoders.end(), encoder);
+  if (it == klass->json_encoders.end()) {
+    klass->json_encoders.push_front(encoder);
+  }
+
+  klass->members.for_each([klass, encoder](ClassMemberInstanceField &field) {
+    std::unordered_set<ClassPtr> sub_classes;
+    field.var->tinf_node.get_type()->get_all_class_types_inside(sub_classes);
+    for (auto sub_klass : sub_classes) {
+      if (sub_klass != klass) {
+        store_json_encoder(sub_klass, encoder);
+      }
+    }
+  });
+}
+
+ClassPtr extract_json_encoder_from_call(VertexAdaptor<op_func_call> call) noexcept {
+  auto first_arg = call->args().front();
+  kphp_assert_msg(first_arg, "absent first arg");
+  auto first_arg_var = first_arg.try_as<op_var>();
+  kphp_assert_msg(first_arg_var, "first arg shoule be op_var type");
+  auto v_encoder = first_arg_var->var_id->init_val.try_as<op_string>();
+  kphp_assert_msg(v_encoder, "absent init value");
+  auto encoder = G->get_class(v_encoder->str_val);
+  kphp_assert_msg(encoder, "couldn't finc encoder class");
+  return encoder;
+}
+
+void check_to_json_impl_call(VertexAdaptor<op_func_call> call) noexcept {
+  kphp_error_return(call->args().size() >= 2, "incorrect args count in to_json_impl call");
+  const auto *type = tinf::get_type(call->args()[1]);
+  kphp_error_return(type->ptype() == tp_Class, "second argument of to_json_impl should be a class type");
+  const auto klass = type->class_type();
+  kphp_error(!klass->is_ffi_cdata(), "called to_json_impl with CData");
+  check_json_fields_duplication(klass);
+  check_json_tag_correctness(klass);
```

Comment:
You have closed the discussion, but it's needed to be reopened again. This place is wrong for checking a class.

1. You check the class A not once, but on each call for encode/decode for A
2. If there are some errors (like names duplication), you get multiple errors instead of one
3. If A is nested, you don't even check it.

Here's what I mean by (3):

```
class Nested {
    /** @kphp-json rename=a */
    public $f1 = 1;
    /** @kphp-json rename=a */
    public $f2 = 2;
}

class Response {
    public Nested $n;
}

$r = new Response;
$r->n = new Nested;
JsonEncoder::encode($r);
// {"n":{"a":1,"a":2}}
```

This should have failed because of fields duplication, but class Nested wasn't checked, because it wasn't directly passed to encode/decode.

Another example: I've mentioned in another comment, that for tuples and others, that are unsupported in json, a compilation error should be fired. If you insert `check_json_fields_types(klass)` here, and class Nested will have tuple, that check won't execute for Nested, and compilation will fail on g++.

A correct way is to perform these checks somewhere around where you add a new encoder for `klass->json_encoders`. Then it will be done once per class and all nested classes would be processed also.

Reply:
I closed previous discussion by a reason. Namely, your concern was about performance, but suggested changes failed to work. So, I preferred to leave as is.
But now you find good problem, which definitely worth to be fixed.
Fixed.